### PR TITLE
[Ref][Core][Opset13] BitwiseAnd, BitwiseOr and BitwiseXor core shell and reference

### PR DESCRIPTION
**Pull request by mmikolajcz**
### Details:
 - *BitwiseAnd-13 core shell and reference implementation*
 - *BitwiseOr-13 core shell and reference implementation*
 - *BitwiseXor-13 core shell and reference implementation*
 - *Support for 1bit and 4bit integers will be considered in separate PR*

PR with spec: https://github.com/openvinotoolkit/openvino/pull/19798

Related to BitwiseNot: https://github.com/openvinotoolkit/openvino/pull/19956

### Tickets:
 - *BitwiseAnd:* 118964, 118963, 118965
 - *BitwiseOr:* 118972, 118971, 118973
 - *BitwiseXor:* 118948, 118947, 118949

### CODE REVIEW

File: src/core/reference/include/openvino/reference/bitwise_and.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * @param broadcast_spec  Broadcast specification mode.
+ */
+template <class T>
+void bitwise_and(const T* arg0,
+                 const T* arg1,
+                 T* out,
+                 const Shape& arg0_shape,
+                 const Shape& arg1_shape,
+                 const op::AutoBroadcastSpec& broadcast_spec) {
+    autobroadcast_binop(arg0, arg1, out, arg0_shape, arg1_shape, broadcast_spec, func::bitwise_and<T>);
```
Comment (praasz) [inline src/core/reference/include/openvino/reference/bitwise_and.hpp:43]:
```suggestion
    autobroadcast_binop(arg0, arg1, out, arg0_shape, arg1_shape, broadcast_spec,
    std::is_same<typename std::decay<T>::type, char>::value ? std::bit_and<bool> : std::bit_and<T>);
```
Try use bitwise functors from <functional> instated custom.
Apply similar to others references.

File: src/core/src/op/bitwise_xor.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+namespace v13 {
+BitwiseXor::BitwiseXor(const Output<Node>& arg0, const Output<Node>& arg1, const AutoBroadcastSpec& auto_broadcast)
+    : BinaryElementwiseBitwise(arg0, arg1, auto_broadcast) {
+    constructor_validate_and_infer_types();
+}
+
+std::shared_ptr<Node> BitwiseXor::clone_with_new_inputs(const OutputVector& new_args) const {
+    OV_OP_SCOPE(v13_BitwiseXor_clone_with_new_inputs);
+    check_new_args_count(this, new_args);
+    return std::make_shared<BitwiseXor>(new_args.at(0), new_args.at(1), this->get_autob());
```
Comment (praasz) [inline src/core/src/op/bitwise_xor.cpp:22]:
```suggestion
    return std::make_shared<BitwiseXor>(new_args.at(0), new_args.at(1), get_autob());
```

File: src/core/tests/type_prop/bitwise_ops.hpp [hunk 3]
```diff
@@ -0,0 +1,1020 @@
...
+#include "common_test_utils/test_assertions.hpp"
+#include "common_test_utils/type_prop.hpp"
+#include "openvino/core/dimension_tracker.hpp"
+#include "openvino/op/util/attr_types.hpp"
+
+using namespace ov;
+using namespace testing;
+
+template <class T>
+class BitwiseOperator : public testing::Test {};
```
Comment (praasz) [inline src/core/tests/type_prop/bitwise_ops.hpp:18]:
```suggestion
class BitwiseOperator : public TypePropOpTest<T> {};
```
Then `make_op` helper can be used  (`common_test_utils/type_prop.hpp`).

File: src/core/tests/type_prop/bitwise_ops.hpp [hunk 4]
```diff
@@ -0,0 +1,1020 @@
...
+
+TYPED_TEST_SUITE_P(BitwiseOperator);
+
+TYPED_TEST_P(BitwiseOperator, default_constructor_integer) {
+    auto A = std::make_shared<ov::op::v0::Parameter>(element::i32,
```
Comment (praasz) [inline src/core/tests/type_prop/bitwise_ops.hpp:23]:
```suggestion
    auto a = std::make_shared<ov::op::v0::Parameter>(element::i32,
```
Usually variables have `lower_case` convention. Unify naming for all variables here.

File: src/core/tests/type_prop/bitwise_ops.hpp [hunk 5]
```diff
@@ -0,0 +1,1020 @@
...
+                                                     PartialShape{-1, 4, 1, 6, Dimension(1, 6), Dimension(2, 6)});
```
Comment (praasz) [inline src/core/tests/type_prop/bitwise_ops.hpp:24]:
```suggestion
                                                     PartialShape{-1, 4, 1, 6, {1, 6}, {2, 6}});
```
Consider to unify creation of partial shape same way in this test file.

File: src/core/tests/type_prop/bitwise_ops.hpp [hunk 6]
```diff
@@ -0,0 +1,1020 @@
...
+    EXPECT_EQ(op->get_element_type(), ov::element::boolean);
+    EXPECT_EQ(op->get_output_partial_shape(0),
+              (ov::PartialShape{-1, 4, 5, 6, ov::Dimension(5, 8), ov::Dimension(5, 6)}));
+}
+
+TYPED_TEST_P(BitwiseOperator, shape_inference_2D) {
+    auto A = std::make_shared<ov::op::v0::Parameter>(element::i32, Shape{2, 2});
+    auto B = std::make_shared<ov::op::v0::Parameter>(element::i32, Shape{2, 2});
+
+    const auto op = std::make_shared<TypeParam>(A, B);
```
Comment (praasz) [inline src/core/tests/type_prop/bitwise_ops.hpp:80]: If test fixuture will be child of `TypePropOpTest` then
```suggestion
    const auto op = this->make_op(a, b);
```
Consider refactor test to use it.

File: src/core/tests/type_prop/bitwise_ops.hpp [hunk 7]
```diff
@@ -0,0 +1,1020 @@
...
+
+#include <vector>
+
+#include "common_test_utils/test_assertions.hpp"
+#include "common_test_utils/type_prop.hpp"
+#include "openvino/core/dimension_tracker.hpp"
+#include "openvino/op/util/attr_types.hpp"
+
+using namespace ov;
+using namespace testing;
```
Comment (praasz) [inline src/core/tests/type_prop/bitwise_ops.hpp:15]: Try to remove it or be consistent in using the namespace as in code is often used `ov::....`

Consider remove it and add required using like
`using ov::op::v0::Parameter` then in test use just `Parameter` etc.

File: src/core/tests/type_prop/bitwise_ops.hpp [hunk 8]
```diff
@@ -0,0 +1,1020 @@
...
+    ov::PartialShape pshape_A{ov::Dimension(-1), ov::Dimension(3), ov::Dimension(2, 224), ov::Dimension(1, 128)};
+    ov::PartialShape pshape_B{ov::Dimension(-1), ov::Dimension(3), ov::Dimension(2, 224), ov::Dimension(1, 128)};
+    ov::PartialShape expected_shape{-1, 3, ov::Dimension(2, 224), ov::Dimension(1, 128)};
+
+    // Equal labels
+    set_shape_labels(pshape_A, {30, 31, 32, 33});
+    set_shape_labels(pshape_B, {30, 31, 32, 33});
+    set_shape_labels(expected_shape, {30, 31, 32, 33});
+
+    auto param_A = std::make_shared<ov::op::v0::Parameter>(element::i32, pshape_A);
```
Comment (praasz) [inline src/core/tests/type_prop/bitwise_ops.hpp:954]:
```suggestion
    auto a = std::make_shared<ov::op::v0::Parameter>(element::i32, pshape_A);
```
In other test this is `A`, Select one naming convention.

File: src/core/include/openvino/op/util/binary_elementwise_bitwise.hpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    BinaryElementwiseBitwise(const Output<Node>& arg0,
+                             const Output<Node>& arg1,
+                             const AutoBroadcastSpec& autob = AutoBroadcastSpec());
+
+public:
+    OPENVINO_OP("BinaryElementwiseBitwise", "util");
+
+    void validate_and_infer_types() override;
+
+    const AutoBroadcastSpec& get_autob() const override {
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/include/openvino/op/util/binary_elementwise_bitwise.hpp]: Done

File: src/core/src/op/bitwise_and.cpp [hunk 10]
```diff
@@ -0,0 +1,27 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#include "openvino/op/bitwise_and.hpp"
+
+#include "itt.hpp"
+#include "openvino/core/validation_util.hpp"
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/src/op/bitwise_and.cpp]: Not required, removed

File: src/core/src/op/bitwise_or.cpp [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+namespace v13 {
+BitwiseOr::BitwiseOr(const Output<Node>& arg0, const Output<Node>& arg1, const AutoBroadcastSpec& auto_broadcast)
+    : BinaryElementwiseBitwise(arg0, arg1, auto_broadcast) {
+    constructor_validate_and_infer_types();
+}
+
+std::shared_ptr<Node> BitwiseOr::clone_with_new_inputs(const OutputVector& new_args) const {
+    OV_OP_SCOPE(v13_BitwiseOr_clone_with_new_inputs);
+    check_new_args_count(this, new_args);
+    return std::make_shared<BitwiseOr>(new_args.at(0), new_args.at(1), this->get_autob());
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/src/op/bitwise_or.cpp]: Done

File: src/core/src/op/util/binary_elementwise_bitwise.cpp [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+                                                                 const Output<Node>& arg1,
+                                                                 const AutoBroadcastSpec& autob)
+    : Op({arg0, arg1}),
+      m_autob(autob) {}
+
+void ov::op::util::BinaryElementwiseBitwise::validate_and_infer_types() {
+    OV_OP_SCOPE(v0_util_BinaryElementwiseBitwise_validate_and_infer_types);
+    auto args_et_pshape = op::util::validate_and_infer_elementwise_args(this);
+    element::Type& args_et = std::get<0>(args_et_pshape);
+    PartialShape& args_pshape = std::get<1>(args_et_pshape);
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/src/op/util/binary_elementwise_bitwise.cpp]: Done

File: src/core/src/op/util/binary_elementwise_bitwise.cpp [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+ov::op::util::BinaryElementwiseBitwise::BinaryElementwiseBitwise(const Output<Node>& arg0,
+                                                                 const Output<Node>& arg1,
+                                                                 const AutoBroadcastSpec& autob)
+    : Op({arg0, arg1}),
+      m_autob(autob) {}
+
+void ov::op::util::BinaryElementwiseBitwise::validate_and_infer_types() {
+    OV_OP_SCOPE(v0_util_BinaryElementwiseBitwise_validate_and_infer_types);
+    auto args_et_pshape = op::util::validate_and_infer_elementwise_args(this);
+    element::Type& args_et = std::get<0>(args_et_pshape);
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/src/op/util/binary_elementwise_bitwise.cpp]: Done

File: src/plugins/template/backend/ops/bitwise_and.cpp [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+template <element::Type_t ET>
+bool evaluate(const std::shared_ptr<ov::op::v13::BitwiseAnd>& node,
+              ov::TensorVector& outputs,
+              const ov::TensorVector& inputs) {
+    OPENVINO_ASSERT(inputs.size() == 2);
+    OPENVINO_ASSERT(outputs.size() == 1);
+    outputs[0].set_shape(infer_broadcast_shape(node.get(), inputs[0].get_shape(), inputs[1].get_shape()));
+    using T = typename ov::element_type_traits<ET>::value_type;
+    ov::reference::bitwise_and(inputs[0].data<T>(),
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/plugins/template/backend/ops/bitwise_and.cpp]: Changed for const inputs

File: src/plugins/template/tests/functional/op_reference/bitwise.hpp [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        case BitwiseTypes::BITWISE_OR: {
+            bitwise_op = std::make_shared<ov::op::v13::BitwiseOr>(params_vec[0], params_vec[1]);
+            break;
+        }
+        case BitwiseTypes::BITWISE_XOR: {
+            bitwise_op = std::make_shared<ov::op::v13::BitwiseXor>(params_vec[0], params_vec[1]);
+            break;
+        }
+        default: {
+            throw std::runtime_error("Incorrect type of Bitwise operation");
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/plugins/template/tests/functional/op_reference/bitwise.hpp]: Modified to use EXPECT_TRUE and removed default for switch

File: src/core/reference/include/openvino/reference/bitwise_and.hpp (hunk 1 above)
Comment (mmikolajcz) [reply src/core/reference/include/openvino/reference/bitwise_and.hpp:43]: This solution caused compiler errors, modified it a bit to use std::bit_<and/or/xor>, please look if solution is acceptable :)

File: src/core/src/op/bitwise_xor.cpp (hunk 2 above)
Comment (mmikolajcz) [reply src/core/src/op/bitwise_xor.cpp]: done

File: src/core/tests/type_prop/bitwise_ops.hpp (hunk 3 above)
Comment (mmikolajcz) [reply src/core/tests/type_prop/bitwise_ops.hpp]: Done

File: src/core/tests/type_prop/bitwise_ops.hpp (hunk 4 above)
Comment (mmikolajcz) [reply src/core/tests/type_prop/bitwise_ops.hpp]: Done

File: src/core/tests/type_prop/bitwise_ops.hpp (hunk 5 above)
Comment (mmikolajcz) [reply src/core/tests/type_prop/bitwise_ops.hpp]: Done, changed DImension(...) to just {...}

File: src/core/tests/type_prop/bitwise_ops.hpp (hunk 6 above)
Comment (mmikolajcz) [reply src/core/tests/type_prop/bitwise_ops.hpp]: Done

File: src/core/tests/type_prop/bitwise_ops.hpp (hunk 7 above)
Comment (mmikolajcz) [reply src/core/tests/type_prop/bitwise_ops.hpp]: removed all ov:... from namespace, modified to use just Parameter

File: src/core/tests/type_prop/bitwise_ops.hpp (hunk 8 above)
Comment (mmikolajcz) [reply src/core/tests/type_prop/bitwise_ops.hpp]: Fixed

File: src/tests/functional/plugin/conformance/test_runner/op_conformance_runner/src/op_impl_check/single_op_graph.cpp [hunk 16]
```diff
@@ -1474,6 +1474,24 @@ std::shared_ptr<ov::Model> generateBinaryEltwise(const std::shared_ptr<ov::op::O
...
+        eltwise = std::make_shared<ov::op::v13::BitwiseAnd>(params[0], params[1]);
+    } else if (ov::is_type<ov::op::v13::BitwiseOr>(node)) {
+        eltwise = std::make_shared<ov::op::v13::BitwiseOr>(params[0], params[1]);
+    } else if (ov::is_type<ov::op::v13::BitwiseXor>(node)) {
+        eltwise = std::make_shared<ov::op::v13::BitwiseXor>(params[0], params[1]);
+    } else {
+        return nullptr;
+    }
+    ov::ResultVector results{std::make_shared<ov::op::v0::Result>(eltwise)};
+    return std::make_shared<ov::Model>(results, ov::ParameterVector{params}, "BinaryEltwiseBitwiseGraph");
```
Comment (t-jankowski) [inline src/tests/functional/plugin/conformance/test_runner/op_conformance_runner/src/op_impl_check/single_op_graph.cpp:1492]:
```suggestion
    return std::make_shared<ov::Model>(results, params, "BinaryEltwiseBitwiseGraph");
```

**Review by t-jankowski — CHANGES_REQUESTED** [review]

File: src/core/reference/include/openvino/reference/bitwise_and.hpp [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "openvino/reference/autobroadcast_binop.hpp"
+
+namespace ov {
+namespace reference {
+namespace func {
+// Check for char datatype used by ov::element::boolean
+template <class T, typename std::enable_if<std::is_same<typename std::decay<T>::type, char>::value>::type* = nullptr>
+T bitwise_and(const T arg0, const T arg1) {
+    return static_cast<T>(std::bit_and<bool>()(arg0, arg1));
```
Comment (praasz) [inline src/core/reference/include/openvino/reference/bitwise_and.hpp:18]: On each iteration the functor is created then do calculation and destroyed. Functor should be created once before loop then only its call operator is used.

This should be corrected to create functor before iterations.

File: src/core/reference/include/openvino/reference/bitwise_and.hpp (hunk 1 above)
Comment (praasz) [reply src/core/reference/include/openvino/reference/bitwise_and.hpp:43]: It should be corrected to create functor in this function and pass it not create it on each iteration.

**Review by praasz — APPROVED** [review]

**Review by t-jankowski — APPROVED** [review]

**Pull request merged**